python_ta/debug/snapshot_tracer.py

```diff
@@ -39,37 +53,119 @@ def __init__(self, output_directory: Optional[str] = None, **kwargs) -> None:
             raise ValueError(
                 "Use the output_directory parameter to specify a different output path."
             )
+        self._snapshots = []
         self._snapshot_args = kwargs
         self._snapshot_args["memory_viz_args"] = copy.deepcopy(kwargs.get("memory_viz_args", []))
-        self._snapshot_counts = 0
-        self.output_directory = output_directory if output_directory else "."
+        self.output_directory = os.path.abspath(output_directory if output_directory else ".")
+        self.open_webstepper = open_webstepper
+        self._first_line = float("inf")
 
     def _trace_func(self, frame: types.FrameType, event: str, _arg: Any) -> None:
         """Take a snapshot of the variables in the functions specified in `self.include`"""
-        if event == "line" and frame.f_locals:
-            self._snapshot_args["memory_viz_args"].extend(
-                [
-                    "--output",
-                    os.path.join(
-                        os.path.abspath(self.output_directory),
-                        f"snapshot-{self._snapshot_counts}.svg",
-                    ),
-                ]
+        if self._first_line == float("inf"):
+            self._first_line = frame.f_lineno
+        if event == "line":
+            filename = os.path.join(
+                self.output_directory,
+                f"snapshot-{len(self._snapshots)}.svg",
             )
+            self._snapshot_args["memory_viz_args"].extend(["--output", filename])
+
             snapshot(
                 save=True,
                 **self._snapshot_args,
             )
-            self._snapshot_counts += 1
+
+            line_number = frame.f_lineno - self._first_line + 1
+            self._add_svg_to_map(filename, line_number)
+
+    def _add_svg_to_map(self, filename: str, line: int) -> None:
+        """Add the SVG in filename to self._snapshots"""
+        with open(filename) as svg_file:
+            svg_content = svg_file.read()
+            self._snapshots.append(
+                {
+                    "lineNumber": line,
+                    "svg": svg_content,
+                }
+            )
 
     def __enter__(self):
         """Set up the trace function to take snapshots at each line of code."""
-        func_frame = inspect.getouterframes(inspect.currentframe())[1].frame
-        func_frame.f_trace = self._trace_func
+        self._func_frame = inspect.getouterframes(inspect.currentframe())[1].frame
```

Comment:
This appears to be a new instance attribute; in general, try to be conservative when adding attributes. As far as I can tell, the reason you've added this is so that `_get_code` can access it. However, both the `__enter__` and `__exit__` methods already had existing code for accessing the correct frame. Preserve that code, and then pass the frame object to `_get_code`